File: backend/skills/tool_manage.go
package skills

import (
	"encoding/json"
	"fmt"

	"github.com/enough/enough/backend/approval"
)

type SkillManageOptions struct {
	GuardEnabled       bool
	MarkCreatedAsAgent bool
	// ArchiveOnDelete turns 'delete' into an archive (move to .archive/).
	// Set for background-review and curator forks — autonomous passes must
	// never destroy data; archives are recoverable. It also hard-protects
	// the curator-protected builtins from autonomous removal.
	ArchiveOnDelete bool
	WriteApproval   bool
	BypassGate      bool
	Origin          string
}

type skillManageArgs struct {
	Action      string `json:"action"`
	Name        string `json:"name"`
	Content     string `json:"content"`
	OldString   string `json:"old_string"`
	NewString   string `json:"new_string"`
	ReplaceAll  bool   `json:"replace_all"`
	Category    string `json:"category"`
	FilePath    string `json:"file_path"`
	FileContent string `json:"file_content"`
	AbsorbedInto string `json:"absorbed_into"`
}

func ExecuteSkillManage(argsJSON string, opts SkillManageOptions) (string, bool) {
	var args skillManageArgs
	_ = json.Unmarshal([]byte(argsJSON), &args)

	var result SkillManageResult
	var err error

	if opts.WriteApproval && !opts.BypassGate {
		payload := map[string]interface{}{
			"action":        args.Action,
			"name":          args.Name,
			"content":       args.Content,
			"old_string":    args.OldString,
			"new_string":    args.NewString,
			"replace_all":   args.ReplaceAll,
			"category":      args.Category,
			"file_path":     args.FilePath,
			"file_content":  args.FileContent,
			"absorbed_into": args.AbsorbedInto,
		}

		gist := approval.SkillGist(args.Action, args.Name, args.Content, args.FilePath, args.OldString, args.NewString)
		origin := opts.Origin
		if origin == "" {
			origin = "agent"
		}

		record, err := approval.StageWrite(approval.SubsystemSkills, payload, gist, origin)
		if err != nil {
			result = SkillManageResult{
				Success: false,
				Error:   "Staging failed: " + err.Error(),
			}
			outBytes, _ := json.MarshalIndent(result, "", "  ")
			return string(outBytes), true
		}

		result = SkillManageResult{
			Success:   true,
			Staged:    true,
			PendingID: record.ID,
			Gist:      gist,
			Message:   fmt.Sprintf("Staged for approval (skills.write_approval is on). Not yet saved — review with /skills pending."),
		}
		outBytes, _ := json.MarshalIndent(result, "", "  ")
		return string(outBytes), false
	}

	switch args.Action {
	case "create":
		result, err = createSkill(args.Name, args.Content, args.Category, opts.GuardEnabled)
	case "edit":
		result, err = editSkill(args.Name, args.Content, opts.GuardEnabled)
	case "patch":
		result, err = patchSkill(args.Name, args.OldString, args.NewString, args.FilePath, args.ReplaceAll, opts.GuardEnabled)
	case "delete":
		if opts.ArchiveOnDelete {
			result, err = archiveDeleteSkill(args.Name, args.AbsorbedInto)
		} else {
			result, err = deleteSkill(args.Name, args.AbsorbedInto, opts.GuardEnabled)
		}
	case "write_file":
		result, err = writeSkillFile(args.Name, args.FilePath, args.FileContent, opts.GuardEnabled)
	case "remove_file":
		result, err = removeSkillFile(args.Name, args.FilePath)
	default:
		result = SkillManageResult{
			Success: false,
			Error:   "Unknown action '" + args.Action + "'. Use: create, edit, patch, delete, write_file, remove_file",
		}
	}

	if err != nil {
		result = SkillManageResult{
			Success: false,
			Error:   err.Error(),
		}
	}

	if result.Success {
		switch args.Action {
		case "create":
			if opts.MarkCreatedAsAgent {
				MarkAgentCreated(args.Name)
			}
		case "patch", "edit", "write_file", "remove_file":
			BumpPatch(args.Name)
		case "delete":
			// Archived skills keep their usage record (state=archived was
			// set by ArchiveSkill); hard deletes forget it.
			if !opts.ArchiveOnDelete {
				Forget(args.Name)
			}
		}
	}

	outBytes, marshalErr := json.MarshalIndent(result, "", "  ")
	if marshalErr != nil {
		return `{"success": false, "error": "json marshal error"}`, true
	}
	return string(outBytes), !result.Success
}

func ApplySkillPending(payload map[string]interface{}, opts SkillManageOptions) (SkillManageResult, error) {
	argsBytes, err := json.Marshal(payload)
	if err != nil {
		return SkillManageResult{Success: false, Error: err.Error()}, err
	}
	opts.BypassGate = true
	resJSON, isErr := ExecuteSkillManage(string(argsBytes), opts)
	var res SkillManageResult
	if err := json.Unmarshal([]byte(resJSON), &res); err != nil {
		return SkillManageResult{Success: false, Error: err.Error()}, err
	}
	if isErr && res.Error == "" {
		res.Error = "Action failed"
	}
	return res, nil
}

File: backend/skills/tool_manage.ts
// PORT: backend/skills/tool_manage.go

import { Effect } from "effect";
import { type runtime } from "../config/config";
import { type SkillManageResult } from "./types";
import { subsystem_skills, stage_write, skill_gist } from "../approval/write_approval";
import {
  createSkill,
  editSkill,
  patchSkill,
  archiveDeleteSkill,
  deleteSkill,
  writeSkillFile,
  removeSkillFile,
} from "./manage_actions";
import { MarkAgentCreated, BumpPatch, Forget } from "./usage";

export interface SkillManageOptions {
  GuardEnabled: boolean;
  MarkCreatedAsAgent: boolean;
  // ArchiveOnDelete turns 'delete' into an archive (move to .archive/).
  // Set for background-review and curator forks — autonomous passes must
  // never destroy data; archives are recoverable. It also hard-protects
  // the curator-protected builtins from autonomous removal.
  ArchiveOnDelete: boolean;
  WriteApproval: boolean;
  BypassGate: boolean;
  Origin: string;
}

export interface SkillManageArgs {
  action: string;
  name: string;
  content?: string;
  old_string?: string;
  new_string?: string;
  replace_all?: boolean;
  category?: string;
  file_path?: string;
  file_content?: string;
  absorbed_into?: string;
}

export function ExecuteSkillManage(
  argsJSON: string,
  opts: SkillManageOptions
): Effect.Effect<[string, boolean], never> {
  let args: SkillManageArgs = { action: "", name: "" };
  try {
    args = JSON.parse(argsJSON);
  } catch {}

  if (opts.WriteApproval && !opts.BypassGate) {
    const payload: Record<string, unknown> = {
      action: args.action,
      name: args.name,
      content: args.content || "",
      old_string: args.old_string || "",
      new_string: args.new_string || "",
      replace_all: !!args.replace_all,
      category: args.category || "",
      file_path: args.file_path || "",
      file_content: args.file_content || "",
      absorbed_into: args.absorbed_into || "",
    };

    const gist = skill_gist(
      args.action,
      args.name,
      args.content || "",
      args.file_path || "",
      args.old_string || "",
      args.new_string || ""
    );
    const origin = opts.Origin !== "" ? opts.Origin : "agent";

    return stage_write(subsystem_skills, payload, gist, origin).pipe(
      Effect.match({
        onFailure: (err) => {
          const result: SkillManageResult = {
            success: false,
            error: "Staging failed: " + err.reason,
          };
          return [JSON.stringify(result, null, "  "), true] as [string, boolean];
        },
        onSuccess: (record) => {
          const result: SkillManageResult = {
            success: true,
            staged: true,
            pending_id: record.id,
            gist: gist,
            message: `Staged for approval (skills.write_approval is on). Not yet saved — review with /skills pending.`,
          };
          return [JSON.stringify(result, null, "  "), false] as [string, boolean];
        },
      })
    );
  }

  let actionEffect: Effect.Effect<SkillManageResult, Error>;
  switch (args.action) {
    case "create":
      actionEffect = createSkill(args.name, args.content || "", args.category || "", opts.GuardEnabled);
      break;
    case "edit":
      actionEffect = editSkill(args.name, args.content || "", opts.GuardEnabled);
      break;
    case "patch":
      actionEffect = patchSkill(
        args.name,
        args.old_string || "",
        args.new_string || "",
        args.file_path || "",
        !!args.replace_all,
        opts.GuardEnabled
      );
      break;
    case "delete":
      if (opts.ArchiveOnDelete) {
        actionEffect = archiveDeleteSkill(args.name, args.absorbed_into || "");
      } else {
        actionEffect = deleteSkill(args.name, args.absorbed_into || "", opts.GuardEnabled);
      }
      break;
    case "write_file":
      actionEffect = writeSkillFile(args.name, args.file_path || "", args.file_content || "", opts.GuardEnabled);
      break;
    case "remove_file":
      actionEffect = removeSkillFile(args.name, args.file_path || "");
      break;
    default:
      actionEffect = Effect.fail(
        new Error(`Unknown action '${args.action}'. Use: create, edit, patch, delete, write_file, remove_file`)
      );
  }

  return actionEffect.pipe(
    Effect.flatMap((result) => {
      if (result.success) {
        switch (args.action) {
          case "create":
            if (opts.MarkCreatedAsAgent) {
              MarkAgentCreated(args.name);
            }
            break;
          case "patch":
          case "edit":
          case "write_file":
          case "remove_file":
            BumpPatch(args.name);
            break;
          case "delete":
            if (!opts.ArchiveOnDelete) {
              Forget(args.name);
            }
            break;
        }
      }
      return Effect.succeed([JSON.stringify(result, null, "  "), !result.success] as [string, boolean]);
    }),
    Effect.catchAll((err) => {
      const result: SkillManageResult = {
        success: false,
        error: err.message,
      };
      return Effect.succeed([JSON.stringify(result, null, "  "), true] as [string, boolean]);
    })
  );
}

export function ApplySkillPending(
  payload: Record<string, unknown>,
  opts: SkillManageOptions
): Effect.Effect<SkillManageResult, Error> {
  return Effect.try({
    try: () => {
      const payloadString = JSON.stringify(payload);
      opts.BypassGate = true;
      // We run ExecuteSkillManage using Effect.runSync because ApplySkillPending in Go
      // calls ExecuteSkillManage synchronously.
      // Since ExecuteSkillManage returns Effect.Effect<[string, boolean], never>, it is safe.
      const [resJSON, isErr] = Effect.runSync(ExecuteSkillManage(payloadString, opts));
      const res = JSON.parse(resJSON) as SkillManageResult;
      if (isErr && !res.error) {
        res.error = "Action failed";
      }
      return res;
    },
    catch: (cause) => (cause instanceof Error ? cause : new Error(String(cause))),
  });
}

/*
PORT STATUS
source path: backend/skills/tool_manage.go
source lines: 156
draft lines: 181
confidence: high
status: phase_b_compile
*/
